Dynamics batching documentation (#3183)
NVIDIA/cuda-quantum: docs/sphinx/using/dynamics.rst

        *   - `runge_kutta`
            - 1st-order (Euler method), 2nd-order (Midpoint method), and 4th-order (classical Runge-Kutta method).

Batch simulation
^^^^^^^^^^^^^^^^^

.. _cudensitymat_batching:

CUDA-Q ``dynamics`` target supports batch simulation, which allows users to run multiple simulations simultaneously.
This batching capability applies to (1) multiple initial states and/or (2) multiple Hamiltonians.

> The batching capability can significantly improve performance when simulating many small identical system dynamics, e.g., parameter sweeping, tomography. 

Reviewer: ```suggestion
Batching can significantly improve performance when simulating many small identical system dynamics, e.g., parameter sweeping or tomography. 
```

For example, we can simulate the time evolution of multiple initial states with the same Hamiltonian as follows:

.. tab:: Python

    .. literalinclude:: ../snippets/python/using/backends/dynamics_state_batching.py
        :language: python
        :start-after: [Begin State Batching]
        :end-before: [End State Batching]

.. tab:: C++

    .. literalinclude:: ../snippets/cpp/using/backends/dynamics_state_batching.cpp
        :language: cpp
        :start-after: [Begin State Batching]
        :end-before: [End State Batching]

> Similarly, we can also batch simulate the time evolution of multiple Hamiltonians as follows:

Reviewer: Can we show the output of the code and specifically that we get results for each initial state (unlike the simple evolve case). Same comment for all code in this section - show the output

Author: Definitely. I've added a code snippet to demonstrate how the batch result can be unpacked, and a simplified sample of the data.

.. tab:: Python

    .. literalinclude:: ../snippets/python/using/backends/dynamics_operator_batching.py
        :language: python
        :start-after: [Begin Operator Batching]
        :end-before: [End Operator Batching]

>     In this example, we show the most generic batching capability, where we each Hamiltonian in the batch corresponds to a specific initial state.

Reviewer: ```suggestion
    In this example, we show the most generic batching capability, where each Hamiltonian in the batch corresponds to a specific initial state.
```

    In other words, the vector of Hamiltonians and the vector of initial states are of the same length.
    If only one initial state is provided, it will be used for all Hamiltonians in the batch.

.. tab:: C++

    .. literalinclude:: ../snippets/cpp/using/backends/dynamics_operator_batching.cpp
        :language: cpp
        :start-after: [Begin Operator Batching]
        :end-before: [End Operator Batching]

Collapse operators and super-operators can also be batched in a similar manner. 
Specifically, if the `collapse_operators` parameter is a nested list of operators (:math:`\{\{L\}_1, \{\{L\}_2, ...\}`), 
then each set of collapsed operators in the list will be applied to the corresponding Hamiltonian in the batch.

> .. note::

Reviewer: I would move it out of a note to regular text

    In order for all Hamiltonians to be batched, they must have the same structure, i.e., same number of product terms and those terms must act on the same degrees of freedom.
    The order of the terms in the Hamiltonian does not matter, nor do the coefficient values/callback functions and the specific operators on those product terms.

>     Here are a couple of examples of Hamiltonians that can or cannot be batched:

Reviewer: That's really helpful :)

            - Second Hamiltonian
            - Batchable?
        *   - :math:`H_1 = \omega_1 \sigma_z(0)`
            - :math:`H_2 = \omega_2 \sigma_z(0)` 
            - Yes (different coefficients, same operator)
        *   - :math:`H_1 = \omega_z \sigma_z(0) + \cos(\omega_xt) \sigma_x(1)`
            - :math:`H_2 = \omega_z \sigma_z(0) + \sin(\omega_xt)  \sigma_x(1)`
            - Yes (same structure, different callback coefficients)
        *   - :math:`H_1 = \omega_z \sigma_z(0) + \cos(\omega_xt) \sigma_x(1)`
            - :math:`H_2 = \omega_z \sigma_z(0) + \cos(\omega_xt) \sigma_y(1)`
            - Yes (different operators on the same degree of freedom)
        *   - :math:`H_1 = \omega_z \sigma_z(0) + \cos(\omega_xt) \sigma_x(1)`
            - :math:`H_2 = \omega_z \sigma_z(0) + \cos(\omega_xt) \sigma_x(1) + \cos(\omega_yt) \sigma_y(1)`
            - No (different number of product terms)
        *   - :math:`H_1 = \omega_z \sigma_z(0) + \cos(\omega_xt) \sigma_{xx}(0, 1)`
            - :math:`H_2 = \omega_z \sigma_z(0) + \cos(\omega_xt) \sigma_x(0)\sigma_x(1)`
            - No (different structures, two-body operators vs. tensor product of single-body operators)

>     When the Hamiltonians are **not** batchable, CUDA-Q will still run the simulations, but each Hamiltonian will be simulated separately in a sequential manner.

Reviewer: Will the user get any indication or warning that the Hamiltonians could not be batched? It will be good if we have that,

.. note::

    Depending on the number of Hamiltonian operators together with factors such as the integrator, schedule step size, and whether intermediate results are stored, the batch simulation can be memory-intensive.
    If you encounter out-of-memory issues, the `max_batch_size` parameter can be used to limit the number of Hamiltonians that are batched together in one run. 

>     For example, if you set `max_batch_size=2`, then we will run the simulations in batches of 2 Hamiltonians at a time, i.e., the first two Hamiltonians will be simulated together, then the next two, and so on.

Reviewer: How would someone set that - is that an argument to evolve? An environment variable?
It will be good to show code or link to reference for setting max_batch_size

Author: I've added an example (Python and C++) demonstrating this.

> 

Reviewer: Unrelated to this specific PR but I see that the following section om MGMN has only a Python example, can we add a C++ example too? (I believe the C++ support was added in this release) @sacpis
